src/server/plugins/addFrameOptionsHeader.js

```diff
@@ -14,20 +14,34 @@
  * permissions and limitations under the License.
  */
 
+import fp from 'fastify-plugin';
 import matcher from 'matcher';
+
 import { getCSP } from './csp';
 
-export default function addFrameOptionsHeader(req, res, next) {
-  const referer = req.get('Referer');
+/**
+ * Fastify Plugin that adds frame options into the header
+ * @param {import('fastify').FastifyInstance} fastify Fastify instance
+ * @param {import('fastify').FastifyPluginOptions} _opts plugin options
+ * @param {import('fastify').FastifyPluginCallback} done plugin callback
+ */
+const addFrameOptionsHeader = (fastify, _opts, done) => {
+  fastify.addHook('onRequest', async (request, reply) => {
+    const referer = request.headers.Referer || request.headers.Referrer || '';
+    const frameAncestorDomains = getCSP()['frame-ancestors'] || [];
+    const trimmedReferrer = referer.replace('https://', '');
+    const matchedDomain = frameAncestorDomains.find((domain) => matcher.isMatch(trimmedReferrer, `${domain}/*`)
+    );
 
-  const frameAncestorDomains = getCSP()['frame-ancestors'];
-  const trimmedReferrer = referer && referer.replace('https://', '');
-  const matchedDomain = frameAncestorDomains && frameAncestorDomains.find((domain) => matcher.isMatch(trimmedReferrer, `${domain}/*`)
-  );
+    if (matchedDomain) {
+      reply.header('X-Frame-Options', `ALLOW-FROM ${referer}`);
+    }
+  });
 
-  if (matchedDomain) {
-    res.set('X-Frame-Options', `ALLOW-FROM ${referer}`);
-  }
+  done();
```

Comment:
why do we call done here instead of next as previously done?

Reply:
this is a fastify naming convention (see https://www.fastify.io/docs/latest/Reference/Plugins/), of course we can call it `next` if we want, but it's better to follow fastify's naming convention to keep everything more fastify-like and make it easier to read and associate.

you'll also see that i have renamed `req` -> `request` and `res` -> `reply`, same idea

Reply:
yea, I just thought that next and done where different in Fastify, good to know

Reply:
in this case, there isn't an equivalent in express since a plugin does more things than an middleware and a middleware gets executed on a request, but yeah they're similar